File: Problem2.py
'''

Time Complexity : O(slen+plen)
Space Complexity : O(slen+plen)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No issues faced

Your code here along with comments explaining your approach

We use two dictionaries, one for pcount. To keep track of elements and the other one to add and remove elements as we traverse thru the
s string. We compare both the dictionaries to see if they're equal and add the respective index

'''

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        plen = len(p)
        slen = len(s)

        if plen > slen: return []

        pCount = defaultdict(int)
        sCount = defaultdict(int)

        for i in range(plen):
            pCount[p[i]] = 1 + pCount.get(p[i],0)
            sCount[s[i]] = 1 + sCount.get(s[i],0)

        res = [0] if pCount == sCount else []

        j = 0
        for i in range(plen,slen):
            sCount[s[i]] = 1 + sCount.get(s[i],0)
            sCount[s[j]] -= 1

            if sCount[s[j]] == 0:
                sCount.pop(s[j])

            j += 1

            if sCount == pCount:
                res.append(j)


        return res

File: problem2.java
//Problem2 - https://leetcode.com/problems/find-the-index-of-the-first-occurrence-in-a-string/

// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this: No

public class problem2 {
    public int strStr(String haystack, String needle) {

        int needleLength = needle.length();
        int haystackLength = haystack.length();

        if (needleLength == 0) {
            return 0;
        }

        int needlePointer = 0;
        int haystackPointer = 0;

        while (haystackPointer < haystackLength) {

            if (haystack.charAt(haystackPointer) == needle.charAt(needlePointer)) {
                needlePointer++;
            } else {
                haystackPointer = haystackPointer - needlePointer;
                needlePointer = 0;
            }

            haystackPointer++;

            if (needlePointer == needleLength) {

                return (haystackPointer - needleLength);

            }

        }

        return -1;

    }
}
